activerecord/lib/active_record/connection_handling.rb

```diff
@@ -135,7 +135,15 @@ def connects_to(database: {}, shards: {})
     #
     # The database kwarg is deprecated and will be removed in 6.2.0 without replacement.
     def connected_to(database: nil, role: nil, shard: nil, prevent_writes: false, &blk)
-      raise NotImplementedError, "`connected_to` can only be called on ActiveRecord::Base" unless self == Base
+      if legacy_connection_handling
+        if self != Base
+          raise NotImplementedError, "`connected_to` can only be called on ActiveRecord::Base with legacy connection handling."
+        end
+      else
+        if self != Base && !abstract_class
+          raise NotImplementedError, "calling `connected_to` with `role` and `shard` is only allowed on ActiveRecord::Base or abstract classes."
```

Comment:
The `role && shard` check went away in the last commit.

```suggestion
          raise NotImplementedError, "calling `connected_to` is only allowed on ActiveRecord::Base or abstract classes."
```